server/storage/wal/wal_test.go

```diff
@@ -1180,3 +1180,203 @@ func TestLastRecordLengthExceedFileEnd(t *testing.T) {
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 3, Data: []byte("b")}, {Index: 4, Data: []byte("c")}, {Index: 5, Data: []byte("c")}},
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 4},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 3, Data: []byte("b")}, {Index: 4, Data: []byte("c")}, {Index: 5, Data: []byte("c")}},
+			},
+		},
+		{
+			name: "before snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			walReadAll: want{
+				wantError: "slice bounds out of range",
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
+		{
+			name: "at snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			readAt: walpb.Snapshot{Index: 3},
+			walReadAll: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
+		{
+			name: "after snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			readAt: walpb.Snapshot{Index: 4},
+			walReadAll: want{
+				wantError: "snapshot not found",
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
```

Comment:
I think we should add a case for `readAllEntries` with wantError when one of the entries having 0 as its Index.

Reply:
Ok, decided to go other direction. Adding test make me notice that wal.ReadAll() supports reading from index 0, not sure that's correct, but we should match the behavior. Changed the function code to now handle index 0. 

Thanks for pointing this out!